sklearn/mixture/_gaussian_mixture_ic.py

```diff
@@ -0,0 +1,820 @@
+                min_val=self.min_components,
+                max_val=self.max_components,
+            )
+            _, labels_init = np.unique(self.label_init, return_inverse=True)
+        else:
+            labels_init = None
+        self.label_init = labels_init
+
+        check_scalar(
+            self.n_init, name="n_init", target_type=int, min_val=1
+        )
+
+        if self.criterion not in ["aic", "bic"]:
+            msg = f'criterion is {self.criterion} but must be "aic" or "bic".'
+            raise ValueError(msg)
+
+        check_scalar(
+            self.max_agglom_size,
+            name="max_agglom_size",
+            target_type=int,
+            min_val=2,
+        )
+
+    def _init_gm_params(self, X, labels, gm_params):
+        labels = labels.reshape(-1, 1)
+        enc = OneHotEncoder()
+        enc.fit(labels)
+        onehot = enc.fit_transform(labels).toarray()
+        weights_init, means_init, precisions_init = _onehot_to_initial_params(
+            X, onehot, gm_params["covariance_type"]
+        )
+        gm_params["weights_init"] = weights_init
+        gm_params["means_init"] = means_init
+        gm_params["precisions_init"] = precisions_init
+        return gm_params
+
+    def _fit_cluster(
+        self, X, X_subset, y, ag_params, gm_params, agg_clustering, seed
+    ):
+        label_init = self.label_init
+        if label_init is not None:
+            gm_params = self._init_gm_params(X, label_init, gm_params)
+        elif ag_params["affinity"] != "none":
+            gm_params = self._init_gm_params(
+                X_subset, agg_clustering, gm_params
+            )
+        else:
+            gm_params["init_params"] = "kmeans"
+        gm_params["reg_covar"] = 0
+        gm_params["max_iter"] = self.max_iter
+        gm_params["random_state"] = seed
+
+        # if none of the iterations converge, bic/aic is set to inf
+        criter = np.inf
+        # below is the regularization scheme
+        while gm_params["reg_covar"] <= 1 and criter == np.inf:
+            model = GaussianMixture(**gm_params)
+            try:
+                # ignoring warning here because if convergence is not reached,
+                # the regularization is automatically increased
+                with warnings.catch_warnings():
+                    warnings.simplefilter("ignore", ConvergenceWarning)
+                    model.fit(X)
+                predictions = model.predict(X)
+                counts = [
+                    sum(predictions == i)
+                    for i in range(gm_params["n_components"])
+                ]
+                # singleton clusters not allowed
+                assert not any([count <= 1 for count in counts])
+
+            except ValueError:
+                gm_params["reg_covar"] = _increase_reg(gm_params["reg_covar"])
+                continue
+            except AssertionError:
+                gm_params["reg_covar"] = _increase_reg(gm_params["reg_covar"])
```

Comment:
As `_fit_cluster` is run in parallel I would highly recommend using a temporary variable here instead of modifying `gm_params["reg_covar"]` in place.

Reply:
I added a temporary variable `reg_covar` here. But got an [AssertionError](https://github.com/scikit-learn/scikit-learn/pull/19562/checks?check_run_id=2511151020) that seemed to be fixed by this [commit](https://github.com/scikit-learn/scikit-learn/pull/19562/commits/37f0df3a5ca7e24feead561e982981d7e2be0e53) inspired by this [code](https://github.com/scikit-learn/scikit-learn/blob/de1262c35e2aa4ee062d050281ee576ce9e35c94/sklearn/inspection/_permutation_importance.py#L26). Does it seem appropriate to you?